### ISSUE TITLE
lint suggestion: unnecessary matching to get borrowed option.

### ISSUE THREAD

**Issue by SamWhited** [rust-lang-nursery/rust-clippy#2270]
A common issue I ran into when I started Rust was taking an option that owns its value and getting an option that has a borrowed value. My naive solution was matching:

```rust
let ownedOption: Option<myType> = whatever();
let borrowedOption: Option<&myType> = match someOption {
  None => None,
  Some(v) => Some(&v),
};
```

Of course, if you're familiar with the Option APIs this can be simplified:

```rust
let ownedOption: Option<myType> = whatever();
let borrowedOption: Option<&myType> = someOption.as_ref();
```

Having clippy detect the unnecessary match and suggest the simplification would be nice.

**Comment by detrumi** [comment]
I'd like to start working on this.
However, the above example isn't working for me. This works:
```Rust
let owned: Option<()> = None;
let borrowed: Option<&()> = match owned {
  None => None,
  Some(ref v) => Some(v),
};
```
But if you do the match like this
```Rust
Some(v) => Some(&v)
```
then the borrowed value `v` [does not live long enough](https://play.rust-lang.org/?gist=e47873bcf64a349aab801399e93c39a8&version=stable).

**Comment by clarcharr** [comment]
Isn't that just using the new match ergonomics? Or am I mistaken?

**Comment by detrumi** [comment]
Enabling the new match ergonomics feature (pattern_binding_modes I think?) does not fix that example, if that's what you're asking.

The new match ergonomics are the reverse of what's happening here.

**Comment by oli-obk** [comment]
fixed in #2281
